src/include/storage/sql_table.h

```diff
@@ -0,0 +1,207 @@
+#pragma once
+#include <list>
+#include <set>
+#include <utility>
+#include <vector>
+#include "catalog/schema.h"
+#include "storage/data_table.h"
+#include "storage/projected_columns.h"
+#include "storage/projected_row.h"
+#include "storage/storage_defs.h"
+
+namespace terrier::storage {
+
+/**
+ * A SqlTable is a thin layer above DataTable that replaces storage layer concepts like BlockLayout with SQL layer
+ * concepts like Schema. This layer will also handle index maintenance, and possibly constraint checking (confirm when
+ * we bring in execution layer). The goal is to hide concepts like col_id_t and BlockLayout above the SqlTable level.
+ * The SqlTable API should only refer to storage concepts via things like Schema and col_oid_t, and then perform the
+ * translation to BlockLayout and col_id_t to talk to the DataTable and other areas of the storage layer.
+ */
+class SqlTable {
+  /**
+   * Contains all of the metadata the SqlTable needs to reference a DataTable. We shouldn't ever have to expose these
+   * concepts to anyone above the SqlTable level. If you find yourself wanting to return BlockLayout or col_id_t above
+   * this layer, consider alternatives.
+   */
+  struct DataTableVersion {
+    DataTable *const data_table;
+    const BlockLayout layout;
+    const ColumnMap column_map;
+  };
+
+ public:
+  /**
+   * Constructs a new SqlTable with the given Schema, using the given BlockStore as the source
+   * of its storage blocks.
+   *
+   * @param store the Block store to use.
+   * @param schema the initial Schema of this SqlTable
+   * @param oid unique identifier for this SqlTable
+   */
+  SqlTable(BlockStore *const store, const catalog::Schema &schema, const table_oid_t oid)
+      : block_store_(store),
+        oid_(oid),
+        table_({new DataTable(block_store_, StorageUtil::BlockLayoutFromSchema(schema).first, layout_version_t(0)),
```

Comment:
I tried that but BlockLayout's default constructor is deleted, so that doesn't work to not initialize it in the initializer list. I can remove the const label on the fields in BlockLayout and then it works.